### PR TITLE
HDFS-15117. EC: Add getECTopologyResultForPolicies to DistributedFileSystem

### PR DESCRIPTION
**Pull request by ayushtkn**
Backport HDFS-15117 along with related Commits
https://issues.apache.org/jira/browse/HDFS-15117

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  16m  0s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +0 :ok: |  prototool  |   0m  0s |  prototool was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 8 new or modified test files.  |
||| _ branch-3.2 Compile Tests _ |
| +0 :ok: |  mvndep  |   3m 16s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  22m 22s |  branch-3.2 passed  |
| +1 :green_heart: |  compile  |   1m 49s |  branch-3.2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 10s |  branch-3.2 passed  |
| +1 :green_heart: |  mvnsite  |   2m 25s |  branch-3.2 passed  |
| +1 :green_heart: |  shadedclient  |  15m 58s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  7s |  branch-3.2 passed  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 39s |  branch-3.2 passed  |
| -0 :warning: |  patch  |   1m 15s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 18s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  cc  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 46s |  the patch passed  |
| -0 :warning: |  checkstyle  |   1m  7s |  hadoop-hdfs-project: The patch generated 5 new + 675 unchanged - 0 fixed = 680 total (was 675)  |
| +1 :green_heart: |  mvnsite  |   2m 11s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  13m 10s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 54s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 52s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 44s |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 114m 41s |  hadoop-hdfs in the patch failed.  |
| +1 :green_heart: |  unit  |  19m 10s |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 39s |  The patch does not generate ASF License warnings.  |
|  |   | 235m 35s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestFSImage |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2261 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc prototool markdownlint |
| uname | Linux 0db58b457b49 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.2 / 970b9a2 |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~16.04-b01 |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/1/artifact/out/diff-checkstyle-hadoop-hdfs-project.txt |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/1/testReport/ |
| Max. process+thread count | 3847 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: hadoop-hdfs-project |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  10m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +0 :ok: |  prototool  |   0m  1s |  prototool was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 8 new or modified test files.  |
||| _ branch-3.2 Compile Tests _ |
| +0 :ok: |  mvndep  |   3m 16s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  22m 23s |  branch-3.2 passed  |
| +1 :green_heart: |  compile  |   1m 54s |  branch-3.2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 10s |  branch-3.2 passed  |
| +1 :green_heart: |  mvnsite  |   2m 27s |  branch-3.2 passed  |
| +1 :green_heart: |  shadedclient  |  16m  7s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 10s |  branch-3.2 passed  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 40s |  branch-3.2 passed  |
| -0 :warning: |  patch  |   1m 15s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 20s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  cc  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 46s |  the patch passed  |
| -0 :warning: |  checkstyle  |   1m  5s |  hadoop-hdfs-project: The patch generated 4 new + 675 unchanged - 0 fixed = 679 total (was 675)  |
| +1 :green_heart: |  mvnsite  |   2m 10s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  13m 11s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 58s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 52s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 45s |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  |  88m 39s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  unit  |  17m 58s |  hadoop-hdfs-rbf in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 40s |  The patch does not generate ASF License warnings.  |
|  |   | 203m 18s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithDFSStriped |
|   | hadoop.fs.contract.router.web.TestRouterWebHDFSContractAppend |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2261 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc prototool markdownlint |
| uname | Linux 5e02ee2787f5 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.2 / 970b9a2 |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~16.04-b01 |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/artifact/out/diff-checkstyle-hadoop-hdfs-project.txt |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs-rbf.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/testReport/ |
| Max. process+thread count | 3887 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: hadoop-hdfs-project |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ayushtkn** [comment]
Test failures aren't related. Checkstyle warnings due to line length, can't be fixed and exist in the original patches as well

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestECAdmin.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  private ECAdmin admin = new ECAdmin(conf);
+
+  private final ByteArrayOutputStream out = new ByteArrayOutputStream();
+  private final ByteArrayOutputStream err = new ByteArrayOutputStream();
+
+  private static final PrintStream OLD_OUT = System.out;
+  private static final PrintStream OLD_ERR = System.err;
+
+  @Rule
+  public Timeout globalTimeout = new Timeout(300000);
```
Comment (ayushtkn) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestECAdmin.java:56]: Changed this :
Was previously -
```
  public Timeout globalTimeout =
      new Timeout(300000, TimeUnit.MILLISECONDS);
```

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolServerSideTranslatorPB.java [hunk 2]
```diff
@@ -1621,6 +1624,24 @@ public UnsetErasureCodingPolicyResponseProto unsetErasureCodingPolicy(
     }
   }
 
+  @Override
+  public GetECTopologyResultForPoliciesResponseProto getECTopologyResultForPolicies(
+      RpcController controller, GetECTopologyResultForPoliciesRequestProto req)
+      throws ServiceException {
+    try {
+      List<String> policies = req.getPoliciesList();
```
Comment (ayushtkn) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolServerSideTranslatorPB.java:1632]: Changed this -
Was previously 
```
      ProtocolStringList policies = req.getPoliciesList();
```
Due to proto compatibility issue

**Review by ayushtkn — COMMENTED** [review]
Have marked the changes made, Rest stays same.

**Review by vinayakumarb — APPROVED** [review]

**Comment by ayushtkn** [comment]
Thsnx @vinayakumarb for the review and the `pro-tip` for merging this through CLI. This being open since a week, got me anxious. Thanx for sparing time on weekend :-) 
As discussed offline, I will rebase the fork and then merge the branch, with last commit having the PR Id by EOD

**Comment by umamaheswararao** [comment]
Oops I missed your review request @ayushtkn . Thanks @vinayakumarb  for review.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  15m 31s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  9s |  https://github.com/apache/hadoop/pull/2261 does not apply to branch-3.2. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2261 |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2261/3/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**